### PR TITLE
Add support for Restream.io stream keys

### PR DESCRIPTION
**Pull request by Surzhynskyi**
Accepting this PR will allow encoders to deliver FTL streams to Restream, which will re-broadcast them to Mixer channels.

Restream stream keys format differs from the Mixer ones, luckily the only changes needed is to remove the first three symbols and allow using underscore as a separator between user id and the stream key.

Disclaimer: I'm not a C guy, so the code is most probably far from being good. Shout out if there is anything that can be improved, I will be happy to do so.

### CODE REVIEW

**Comment by QuinnDamerell-MS** [comment]
Thanks for the pr. I’m trying to understand what exactly this will do, so this change is needed for the sdk to be used in restream’s service to forward a FTL connection along to Mixer’s ingest?

**Comment by Surzhynskyi** [comment]
@QuinnDamerell-MS thanks for finding time for this.

Accepting and forwarding FTL to Mixer is the goal and it's already implemented on our side. The blocker is now on the SDK side because of how FTL stream keys are getting parsed.

SDK is still to be used in encoders (OBS, XSplit, etc.). This change will allow them to stream to Restream, which will deliver the FTL stream to one or multiple Mixer channels as well as any other channels on any streaming platform in RTMP.

Mixer key example: 12345-r4b6afwoz2bl7zdc99q444lv5ke0dkpe
Mixer key format: {userId}-{hash}

Restream key example: re_12345_5fa06930315e18d73616
Restream key format: re_{userId}_{hash}

What this PR does technically:
1. Allows using underscore as a separator between userId and hash.
2. Removes first 3 symbols of the stream key before parsing it when the key starts with "re_".

**Comment by QuinnDamerell-MS** [comment]
Sorry for all of the questions, but I'm wondering, why can't the key be corrected before given to the SDK? I'm not opposed to the change, it's just we don't have any streaming platform specific logic in the SDK right now, so I'm debating in my head if we should allow it or not. If there's a good reason we can add it, but why not just correct the key before it gets called?

**Comment by Surzhynskyi** [comment]
@QuinnDamerell-MS Sure, that's the question I've been asking myself before submitting this PR. 

If keys are to be corrected on encoder level each encoder would have to implement a stream key conversion hack.

On one hand, it doesn't require changes to the SDK, on other, it makes encoders to include some unnecessary code that could be omitted. It may also create some confusion and hard to debug issues on encoder users and developers when stream keys are changed under the hood based on some conditions, while parsing of the userId and hash is a part of FTL authorization, and it's handled by the FTL SDK already.

Long story short: I see encoder-level parsing/reformatting of stream keys as a hack, so I would prefer to avoid having it if we can.

**Comment by Surzhynskyi** [comment]
@QuinnDamerell-MS Any thoughts on the above? Let me know what you think. 

We put a lot of resources into FTL support and we are really looking forward to launching it. If accepting this is not possible we would have to proceed with per-encoder changes or our own fork of the SDK.

**Comment by QuinnDamerell-MS** [comment]
Sorry for the delay. What about the code that integrates the FTL SDK? Can it handle the logic?

**Comment by Surzhynskyi** [comment]
@QuinnDamerell-MS Sorry, I'm not sure I understand the question.
Do you mean the code in each of the encoders: OBS, XSplit, etc?

**Comment by QuinnDamerell-MS** [comment]
Oh, I think I understand what you're saying. What you are expecting is a user to be using OBS, set it up to stream FTL to restream, provide the restream key, and then stream. Internally OBS would then use the FTL SDK to connect to your service and it would send the modified stream key?

I was thinking this change was needed on your server side, to take the stream from the client and restream it to Mixer's ingests.

**Comment by Surzhynskyi** [comment]
@QuinnDamerell-MS Yes, that's exactly what we are going to do, and that's why we want this change in the SDK itself.

OBS(or any other encoder) will use the Restream stream key to stream FTL to Restream. On our server side, we will accept the stream with Restream stream key and deliver FTL to Mixer using the Mixer key.

File: libftl/ftl-sdk.c [hunk 1]
```diff
@@ -417,25 +417,40 @@ FTL_API char* ftl_status_code_to_string(ftl_status_t status) {
 
 BOOL _get_chan_id_and_key(const char *stream_key, uint32_t *chan_id, char *key) {
   size_t len, i = 0;
-  
+
+  char * copy_of_key = _strdup(stream_key);
+
   len = strlen(stream_key);
```
Comment (QuinnDamerell-MS) [inline libftl/ftl-sdk.c:427]: I would add a quick validation to make sure the key != null and that the len is > than some value (at least 3)

File: libftl/ftl-sdk.c [hunk 2]
```diff
@@ -417,25 +417,40 @@ FTL_API char* ftl_status_code_to_string(ftl_status_t status) {
...
+  if (strncmp(stream_key, restream_stream_key_marker, strlen(restream_stream_key_marker)) == 0) {
+    for (i = 0; i < len - 3; i++) {
+      copy_of_key[i] = copy_of_key[i + 3];
+    }
+
+    copy_of_key[i] = '\0';
+
+    len = strlen(copy_of_key);
+  }
+  free(restream_stream_key_marker);
```
Comment (QuinnDamerell-MS) [inline libftl/ftl-sdk.c]: You don't need to free this.

**Comment by Surzhynskyi** [comment]
@QuinnDamerell-MS Please let me know if you have more questions or some requests to the code in this PR. I will be happy to get things fixed/explained.

**Review by QuinnDamerell-MS — CHANGES_REQUESTED** [review]

**Comment by QuinnDamerell-MS** [comment]
Sorry, I didn't realize I didn't commit those changes Friday.

File: libftl/ftl-sdk.c (hunk 2 above)
Comment (Surzhynskyi) [reply libftl/ftl-sdk.c]: @QuinnDamerell-MS Are you sure? Why do we need to free copy_of_key and we don't have to free restream_stream_key_marker? They both are char pointers.

File: libftl/ftl-sdk.c (hunk 1 above)
Comment (Surzhynskyi) [reply libftl/ftl-sdk.c:427]: Validation added

**Comment by Surzhynskyi** [comment]
@QuinnDamerell-MS Thanks for the review. I have fixed the issue with validation and replied to your second notice. Will be happy to fix it as well if it's really needed, as far as I understand every memory allocation should be freed to prevent memory leaks.

File: libftl/ftl-sdk.c (hunk 2 above)
Comment (QuinnDamerell-MS) [reply libftl/ftl-sdk.c]: Yeah. strdup is calling malloc to allocate a new block of memory in the heap for the string. (https://en.cppreference.com/w/c/experimental/dynamic/strdup). Anytime you allocate from the heap you need to tell the OS you're done with the memory by calling free.

Both `copy_of_key` and `restream_stream_key_marker` are pointers allocated on the stack. Anything on the stack can't be a dynamic size, thus the compiler and allocate and free them when it knows it's safe to. (when the function is being called and when the function is being returned from.) However, `restream_stream_key_marker` is pointing to a read-only block of memory where the static string "re_" is stored. Only one instance of "re_" ever exists in memory per process. It's not allocated by this function, there's just a pointer pointing to it. So you don't need to call free on it.

I'm not even sure what calling free on it will do, I think it will crash.

Comment (QuinnDamerell-MS) [reply libftl/ftl-sdk.c]: Actually, here's a good post on it. 
https://stackoverflow.com/questions/9504588/should-i-free-char-initialized-using-string-literals

**Review by QuinnDamerell-MS — CHANGES_REQUESTED** [review]

**Comment by Surzhynskyi** [comment]
@QuinnDamerell-MS Fixed. Thank you for the detailed explanation.

**Comment by QuinnDamerell-MS** [comment]
Looks good! Thanks!

**Comment by Surzhynskyi** [comment]
Thank you for all the help, Quinn.
